File: backend_fastapi/app/schemas/dog.py
"""
Dog Pydantic schemas for the K9 Operations Management System
"""
from pydantic import Field
from typing import Optional, List
from datetime import datetime, date
from uuid import UUID

from .enums import DogStatus, DogGender
from .common import BaseSchema, UUIDMixin, TimestampMixin


class DogBase(BaseSchema):
    """Base Dog schema with common fields"""
    name: str = Field(..., min_length=1, max_length=100)
    code: str = Field(..., min_length=1, max_length=20)
    breed: str = Field(..., min_length=1, max_length=100)
    family_line: Optional[str] = Field(None, max_length=100)
    gender: DogGender
    birth_date: date
    microchip_id: Optional[str] = Field(None, max_length=50)
    current_status: DogStatus = DogStatus.ACTIVE
    location: Optional[str] = Field(None, max_length=100)
    specialization: Optional[str] = Field(None, max_length=100)
    color: Optional[str] = Field(None, max_length=50)
    weight: Optional[float] = Field(None, gt=0)
    height: Optional[float] = Field(None, gt=0)


class DogCreate(DogBase):
    """Schema for creating a new dog"""
    birth_certificate: Optional[str] = Field(None, max_length=255)
    photo: Optional[str] = Field(None, max_length=255)
    medical_records: List[str] = Field(default_factory=list)
    assigned_to_user_id: Optional[UUID] = None
    father_id: Optional[UUID] = None
    mother_id: Optional[UUID] = None


class DogUpdate(BaseSchema):
    """Schema for updating a dog - all fields optional"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    code: Optional[str] = Field(None, min_length=1, max_length=20)
    breed: Optional[str] = Field(None, min_length=1, max_length=100)
    family_line: Optional[str] = Field(None, max_length=100)
    gender: Optional[DogGender] = None
    birth_date: Optional[date] = None
    microchip_id: Optional[str] = Field(None, max_length=50)
    current_status: Optional[DogStatus] = None
    location: Optional[str] = Field(None, max_length=100)
    specialization: Optional[str] = Field(None, max_length=100)
    color: Optional[str] = Field(None, max_length=50)
    weight: Optional[float] = Field(None, gt=0)
    height: Optional[float] = Field(None, gt=0)
    birth_certificate: Optional[str] = Field(None, max_length=255)
    photo: Optional[str] = Field(None, max_length=255)
    medical_records: Optional[List[str]] = None
    assigned_to_user_id: Optional[UUID] = None
    father_id: Optional[UUID] = None
    mother_id: Optional[UUID] = None


class DogInDB(DogBase, UUIDMixin, TimestampMixin):
    """Schema representing Dog in database"""
    birth_certificate: Optional[str] = None
    photo: Optional[str] = None
    medical_records: List[str]
    assigned_to_user_id: Optional[UUID] = None
    father_id: Optional[UUID] = None
    mother_id: Optional[UUID] = None


class DogResponse(DogBase, UUIDMixin):
    """Schema for API responses"""
    birth_certificate: Optional[str] = None
    photo: Optional[str] = None
    medical_records: List[str]
    created_at: datetime
    updated_at: Optional[datetime] = None
    assigned_to_user_id: Optional[UUID] = None
    father_id: Optional[UUID] = None
    mother_id: Optional[UUID] = None

File: k9-frontend/src/types/dog.ts
export const DogStatus = {
  ACTIVE: 'ACTIVE',
  TRAINING: 'TRAINING',
  RETIRED: 'RETIRED',
  INJURED: 'INJURED',
  DECEASED: 'DECEASED',
} as const;

export type DogStatus = typeof DogStatus[keyof typeof DogStatus];

export const DogGender = {
  MALE: 'MALE',
  FEMALE: 'FEMALE',
} as const;

export type DogGender = typeof DogGender[keyof typeof DogGender];

export interface Dog {
  id: string;
  name: string;
  code: string;
  breed: string;
  family_line?: string;
  gender: DogGender;
  birth_date: string;
  microchip_id?: string;
  current_status: DogStatus;
  location?: string;
  specialization?: string;
  color?: string;
  weight?: number;
  height?: number;
  birth_certificate?: string;
  photo?: string;
  medical_records: string[];
  assigned_to_user_id?: string;
  father_id?: string;
  mother_id?: string;
  created_at: string;
  updated_at?: string;
}

export interface DogCreate {
  name: string;
  code: string;
  breed: string;
  family_line?: string;
  gender: DogGender;
  birth_date: string;
  microchip_id?: string;
  current_status?: DogStatus;
  location?: string;
  specialization?: string;
  color?: string;
  weight?: number;
  height?: number;
  birth_certificate?: string;
  photo?: string;
  medical_records?: string[];
  assigned_to_user_id?: string;
  father_id?: string;
  mother_id?: string;
}

export interface DogUpdate {
  name?: string;
  code?: string;
  breed?: string;
  family_line?: string;
  gender?: DogGender;
  birth_date?: string;
  microchip_id?: string;
  current_status?: DogStatus;
  location?: string;
  specialization?: string;
  color?: string;
  weight?: number;
  height?: number;
  birth_certificate?: string;
  photo?: string;
  medical_records?: string[];
  assigned_to_user_id?: string;
  father_id?: string;
  mother_id?: string;
}

export interface DogFilters {
  skip?: number;
  limit?: number;
  status?: DogStatus;
  gender?: DogGender;
  breed?: string;
  search?: string;
}

export interface PaginatedDogs {
  items: Dog[];
  total: number;
  skip: number;
  limit: number;
}

export interface DogStatistics {
  total: number;
  by_status: Record<string, number>;
  by_gender: Record<string, number>;
  by_breed: Record<string, number>;
  top_breeds: [string, number][];
}
